File: Android/VKAudioPlayer/app/src/main/java/github/y0rrrsh/vkaudioplayer/database/syncitem/SyncItemDB.java
package github.y0rrrsh.vkaudioplayer.database.syncitem;

import android.content.Context;

import github.y0rrrsh.vkaudioplayer.database.vkitem.VkItemDB;
import io.realm.Realm;
import io.realm.RealmConfiguration;
import io.realm.RealmResults;

/**
 * @author Artur Yorsh. 14.06.16.
 */
public class SyncItemDB {

    private static SyncItemDB instance;
    private static final String BASE_NAME = "vkap_sync_list_";

    private Realm realm;

    private SyncItemDB(Context context, int userId) {
        RealmConfiguration realmConfiguration = new RealmConfiguration.Builder(context)
                .name(BASE_NAME + userId)
                .deleteRealmIfMigrationNeeded()
                .build();
        realm = Realm.getInstance(realmConfiguration);
    }

    public static void init(Context context, int userId) {
        instance = new SyncItemDB(context, userId);
    }

    public static SyncItemDB getInstance() {
        return instance;
    }

    public boolean isSyncEnabledForId(int id) {
        return realm.where(SyncItem.class).equalTo("id", id).findFirst() != null;
    }

    public void setSyncEnabledForId(int id, boolean enabled) {
        if (enabled) {
            SyncItem item = new SyncItem(id);
            realm.executeTransaction(realm1 -> realm1.copyToRealmOrUpdate(item));
        } else {
            realm.executeTransaction(realm1 -> realm1.where(SyncItem.class).equalTo("id", id).findAll().deleteAllFromRealm());
        }
    }

    public long getSyncMillisForId(int id) {
        SyncItem syncItem = realm.where(SyncItem.class).equalTo("id", id).findFirst();
        return syncItem == null ? 0 : syncItem.getLastSyncMillis();
    }

    public void setSyncMillisForId(int id, long millis) {
        SyncItem item = realm.where(SyncItem.class).equalTo("id", id).findFirst();
        if (item != null) realm.executeTransaction(realm1 -> item.setLastSyncMillis(millis));
    }

    /**
     * Remove all items from this db,
     * that are not contained in user's vk items(groups, or friends could be deleted from website)
     */
    public void update() {
        RealmResults<SyncItem> syncItems = realm.where(SyncItem.class).findAll();
        for (SyncItem syncItem : syncItems) {
            if (!VkItemDB.getInstance().contains(syncItem.getId())) {
                setSyncEnabledForId(syncItem.getId(), false);
            }
        }
    }
}

File: iOS/VkAudio/database/SyncItemDB.swift
//
//  SyncItemDB.swift
//  VkAudio
//
//  Created by mac-224 on 24.06.16.
//  Copyright © 2016 y0rrrsh. All rights reserved.
//

import Foundation
import RealmSwift

class SyncItemDB {
    
    private static let BASE_NAME = "vkap_sync_list"
    static var sharedInstance: SyncItemDB?
    
    private var realm: Realm
    
    private init(userId: Int) {
        var realmConfig = Realm.Configuration()
        realmConfig.deleteRealmIfMigrationNeeded = true
        
        let fileUrl = NSURL.fileURLWithPath(realmConfig.fileURL!.path!)
            .URLByDeletingLastPathComponent!
            .URLByAppendingPathComponent("\(SyncItemDB.BASE_NAME)_\(userId)")
            .URLByAppendingPathExtension("realm")
        
        realmConfig.fileURL = fileUrl
        realm = try! Realm(configuration: realmConfig)
    }
    
    static func setup(userId: Int) {
        sharedInstance = SyncItemDB(userId: userId)
    }
    
    func getAllIds() -> [Int]? {
        let syncItems = realm.objects(SyncItem.self)
        return syncItems.map{syncItem in syncItem.id}
    }
    
    func isSyncEnabledForId(id: Int) -> Bool {
        return realm.objectForPrimaryKey(SyncItem.self, key: id) != nil
    }
    
    func setSyncEnabledForId(id: Int, enabled: Bool) {
        try! realm.write {
            if enabled {
                realm.add(SyncItem(value: ["id": id]), update: true)
            }
            else if let item = realm.objectForPrimaryKey(SyncItem.self, key: id) {
                realm.delete(item)
            }
        }
    }
    
    func getSyncMillisForId(id: Int) -> NSDate {
        if let item = realm.objectForPrimaryKey(SyncItem.self, key: id) {
            return item.lastSyncMillis
        }
        return NSDate(timeIntervalSince1970: 0)
    }
    
    func setSyncMillisForId(id: Int, millis: NSDate) {
        if let item = realm.objectForPrimaryKey(SyncItem.self, key: id) {
            try! realm.write({ 
                item.lastSyncMillis = millis
            })
        }
    }
    
    func update() {
        let items = realm.objects(SyncItem.self)
        for item in items {
            if !VkModelDB.sharedInstance!.contains(item.id) {
                setSyncEnabledForId(item.id, enabled: false)
            }
        }
    }
}